kernel_tuner/nvml.py

```diff
@@ -116,6 +116,7 @@ def set_clocks(self, mem_clock, gr_clock):
             raise ValueError("Graphics clock incompatible with memory clock")
         try:
             pynvml.nvmlDeviceSetApplicationsClocks(self.dev, mem_clock, gr_clock)
+            pynvml.nvmlDeviceSetMemoryLockedClocks(self.dev, mem_clock, mem_clock)
```

Comment:
Nice catch! I didn't know about nvmlDeviceSetMemoryLockedClocks, it seems it overrides the applications clock even. Great!

Reply:
The `nvmlDeviceSetMemoryLockedClocks` was added in [this](https://github.com/gpuopenanalytics/pynvml/commit/30133c72805681a244f13736bc34e440be6503a2#diff-c9c6160b3107b52ed919117b3cba3bd81f8539baa056cf1e8693329a26ba5bc7R2370) commit. Thus, you will need pynvml 11.4.0 or 11.4.1 to use this feature.

It is probably a good idea to add a guard around this statement, such that it is not executed when using an older Python version. There are various [options](https://stackoverflow.com/questions/20180543/how-to-check-version-of-python-modules). For Python >= 3.8:
```
>>> from importlib.metadata import version
>>> version('pynvml')
'11.4.1'
```
For Python < 3.8:
```
>>> import pkg_resources
>>> pkg_resources.get_distribution('pynvml')
pynvml 11.4.1 (/var/software/spack/opt/spack/linux-rocky8-zen2/gcc-9.4.0/py-pynvml-11.4.1-toiwpo2dxodua3rvvbz2efpks6loybgs/lib/python3.9/site-packages)
```
Or with some code specific to pynvml :
```
>>> import pynvml                           
>>> pynvml._version.get_versions()['version']
'11.4.1'
```
What would you prefer?

Reply:
The `nvmlDeviceSetMemoryLockedClocks` call (or `-lmc  --lock-memory-clocks=` in `nvidia-smi`) seems to be only supported on Ampere. With the same driver (495.29.05), setting locked clocks does not work on Turing:
```
# nvidia-smi -lmc 7001,7001
Setting locked Memory clocks is not supported for GPU 00000000:02:00.0.
Treating as warning and moving on.
All done.
```
And neither on Pascal (same error).

I would like to have the `nvmlDeviceSetMemoryLockedClocks` option in kernel_tuner, but it can only be applied when a number of criteria are met. It may be better to require the user to explicitly enable it. Or alternatively, add a try-except block (possible already during initialization) to check whether the capability is available. I am happy to implement such a safeguard, but I do need your feedback on how to proceed.

Reply:
Sorry for the delay in my response! I have been thinking about how to best support this feature. I'm currently leaning towards asking the user to enable it. Properly trying to detect whether the functionality is supported by the GPU at initialization also requires privileges and trying to use those without the user explicitly asking you to doesn't feel good. If you could implement such a safeguard as you proposed that would be great!

With regard to the version of pynvml, I think it would be best to simply ask for at least version 11.4.0 in setup.py:
https://github.com/benvanwerkhoven/kernel_tuner/blob/master/setup.py#L39

Reply:
I added the requirement to `setup.py`. Since pynvml 11.4.1 is now also out, I used that version.